### PR TITLE
Make DisableRequestSuccessLog configurable.

### PR DESCRIPTION
**Pull request by DylanGuedes**
- Make `DisableRequestSuccessLog` configurable by adding a YAML key to its definition
- Modify the http logging middleware to not log successful requests if `DisableRequestSuccessLog` is truthy. This is especially useful to avoid excessive logging for servers that receive too many successful requests without forcing the logger to change from debug level.

### CODE REVIEW

File: middleware/logging.go [hunk 1]
```diff
@@ -94,6 +96,11 @@ func (l Log) Wrap(next http.Handler) http.Handler {
 
 			return
 		}
+
+		if statusCode >= 200 && statusCode < 300 && l.DisableRequestSuccessLog {
```
Comment (bboreham) [inline middleware/logging.go:100]: Would you consider phrasing the three possibilities as `switch` cases?

File: middleware/logging.go [hunk 2]
```diff
@@ -14,11 +14,12 @@ import (
 
 // Log middleware logs http requests
 type Log struct {
-	Log                   logging.Interface
-	LogRequestHeaders     bool // LogRequestHeaders true -> dump http headers at debug log level
-	LogRequestAtInfoLevel bool // LogRequestAtInfoLevel true -> log requests at info log level
-	SourceIPs             *SourceIPExtractor
-	HttpHeadersToExclude  map[string]bool
+	Log                      logging.Interface
+	DisableRequestSuccessLog bool
```
[the remark this answers is not in the document]
Comment (DylanGuedes) [reply middleware/logging.go:18]: Good point. I can definitely rewrite it but I did it this way because:
- It is more consistent with the existing DisableRequestSuccessLog
- It allow `weaveworks/common` users to leave it unset by default (i.e: won't introduce an unwanted change)
That said, I'm happy with either; based on those two points, do you prefer `DisableRequestSuccessLog` (which is false by default) ot `LogSuccesfulRequests` (which is truthy by default)? If you prefer the former, would you like to see a rewrite of  `DisableRequestSuccessLog` too?

File: middleware/logging.go (hunk 1 above)
Comment (DylanGuedes) [reply middleware/logging.go:100]: sure thing.

File: server/server.go [hunk 3]
```diff
@@ -83,7 +83,7 @@ type Config struct {
 
 	RegisterInstrumentation  bool `yaml:"register_instrumentation"`
 	ExcludeRequestInLog      bool `yaml:"-"`
-	DisableRequestSuccessLog bool `yaml:"-"`
+	DisableRequestSuccessLog bool `yaml:"disable_request_success_log"`
```
Comment (bboreham) [inline server/server.go:86]: Somehow my comment here got lost. 
I think if you do this without adding a flag it shows up in docs with no explanation, which is undesirable.
[Example](https://github.com/grafana/mimir/pull/4922/files#diff-809450d2d139bdb2005dd008d79c44bbad79aa5f76dacd68a5b46c7f3fe669c6R573)
I would prefer you expose in your own program, and leave it unexposed in weaveworks/common, as ExcludeRequestInLog is.

Comment (DylanGuedes) [reply server/server.go:86]: aháaaa I definitely overlooked that, great call. That for sure works, thank you.

File: middleware/logging.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (bboreham) [reply middleware/logging.go:18]: > It is more consistent with the existing DisableRequestSuccessLog

Can you rephrase?  It looks to me like you are adding DisableRequestSuccessLog.

> leave it unset by default (i.e: won't introduce an unwanted change)

This doesn't apply to anyone calling `NewLogMiddleware`; do you think anyone is doing it another way?

[the remark this answers is not in the document]
Comment (DylanGuedes) [reply middleware/logging.go:18]: > > It is more consistent with the existing DisableRequestSuccessLog
> 
> Can you rephrase? It looks to me like you are adding DisableRequestSuccessLog.

I was referring to the existing `DisableRequestSuccessLog` defined under `server/server.go`.

> > leave it unset by default (i.e: won't introduce an unwanted change)
> 
> This doesn't apply to anyone calling `NewLogMiddleware`; do you think anyone is doing it another way?

Hmm I thought that mostly users were only doing `server.New()`, without defining `NewLogMiddleware` directly. I'll check how we do it, but tbh your suggestion of defining a flag on my program will make this work any way I prefer, so indeed I shouldn't be too concerned about it.

[the remark this answers is not in the document]
Comment (bboreham) [reply middleware/logging.go:18]: OK, I didn't know there was already one in `middleware.GRPCServerLog`.
In that case, you can add one for `middleware.Log` and make it configurable in exactly the same way, no extra parameters.

File: server/server.go (hunk 3 above)
Comment (bboreham) [reply server/server.go:86]: Now I understand the context a bit better, my preference is to leave this as `"-"`.
You can add any config or flags you want in your downstream application.

Comment (DylanGuedes) [reply server/server.go]: Addressed by https://github.com/weaveworks/common/pull/284/commits/ed6819b7603030a18fc24bec7dbfbce39f318a05.

File: middleware/logging.go (hunk 1 above)
Comment (DylanGuedes) [reply middleware/logging.go]: Addressed by https://github.com/weaveworks/common/pull/284/commits/565676c80fb17b86d151a6863c6f18517ea0c809. WDYT?

File: middleware/logging.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (DylanGuedes) [reply middleware/logging.go:18]: I'm not sure if that's what you had in mind but I pushed https://github.com/weaveworks/common/pull/284/commits/d9330659bfde5eaaa4e1877f3e56a39ed2dc2648 that does the following:
- Remove `DisableRequestSuccessLog` from the list of parameters, to avoid a breaking change to `NewLogMiddleware` users
- Make `LogMiddleware.DisableRequestSuccessLog` default value to be whatever is set on `server.Config.DisableRequestSuccessLog`.

File: middleware/logging_test.go [hunk 4]
```diff
@@ -57,6 +58,51 @@ func TestBadWriteLogging(t *testing.T) {
...
+		recorder := httptest.NewRecorder()
+
+		w := errorWriter{
+			err: tc.err,
+			w:   recorder,
+		}
+		loggingHandler.ServeHTTP(w, req)
+		content := buf.String()
+
+		require.True(t, strings.Contains(content, tc.logContains))
```
Comment (bboreham) [inline middleware/logging_test.go:102]: This is a bit weak for the second test case.
Wouldn't it be better to require that it does _not_ contain the request?

**Review by bboreham — APPROVED** [review]
Thanks.  I have one more suggestion.

File: middleware/logging_test.go (hunk 4 above)
Comment (DylanGuedes) [reply middleware/logging_test.go]: Done! WDYT?
